src/System.Slices/System/Buffers/ReferenceCounter.cs

```diff
@@ -0,0 +1,203 @@
+﻿using System.Runtime.CompilerServices;
+
+namespace System.Runtime
+{
+    /// <summary>
+    /// Make sure the struct is not copied, i.e. pass it only by reference
+    /// </summary>
+    /// <remarks>
+    /// The  counter is not completly race-free. Reading GetGlobalCount and AddReference/Release are subject to a race.
+    /// </remarks>
+    public struct ReferenceCounter
+    {
+        // thread local counts that can be updated very efficiently
+        [ThreadStatic]
+        static ObjectTable t_threadLocalCounts;
+
+        // all thread local counts; these are tallied up when global count is comupted
+        static ObjectTable[] s_allTables = new ObjectTable[4];
+        static int s_allTablesCount;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void AddReference(object obj)
+        {
+            var localCounts = t_threadLocalCounts;
+            if (localCounts != null)
+            {
+                localCounts.Increment(obj);
+            }
+            else
+            {
+                AddNewObject(obj);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Release(object obj)
+        {
+            var localCounts = t_threadLocalCounts;
+            localCounts.Decrement(obj);
```

Comment:
Yeah, I still don't have a well formed opinion on how much error checking we should be doing in our high performance libraries. The CLR already guards against null here, so what's the point of more guards? The only benefit I see is better exception message, but then I am not sure how much it's worth it. If I do add the check, then I have to play games with the inliner (as it does not like throw). Anyway, as I said I don't have an opinion that is firm here, so I can go either way.